muspinsim/cpp/parallel.cpp

```diff
@@ -0,0 +1,239 @@
+#include "base.h"
+
+/*
+ * Computes V^\dagger (M \otimes 1_{d}) V between a complex vector V and a
+ * matrix M where 1_{d} an identity matrix of size d
+ *
+ * For speed, no checks are performed here. The matrix M should be Hermitian
+ * with size N and the shape of V should be (N*d, 1).
+ *
+ * @param V The vector
+ * @param M The matrix
+ * @param d The dimension of the identity matrix
+ *
+ * @return The result of the combined product. Will be a single value.
+ */
+double parallel::fast_measure(np_array_complex_t& V, np_array_complex_t& M, long int d) {
+    py::buffer_info V_info = V.request();
+    py::buffer_info M_info = M.request();
+
+    return parallel::fast_measure_ptr(
+        static_cast<std::complex<double>*>(V_info.ptr),
+        V_info.shape[0],
+        static_cast<std::complex<double>*>(M_info.ptr),
+        M_info.shape[0],
+        d);
+}
+
+/* Same as above, but only for use in C++ - uses pointers instead of numpy
+ * arrays
+ *
+ * @param V_ptr Pointer to the vector
+ * @param V_dim Dimension of V
+ * @param M_ptr Pointer to the matrix
+ * @param M_dim Dimension of M
+ * @param d The dimension of the identity matrix
+ *
+ * @return The result of the combined product. Will be a single value.
+ */
+double parallel::fast_measure_ptr(std::complex<double>* V_ptr, unsigned int V_dim, std::complex<double>* M_ptr, unsigned int M_dim, long int d) {
+    double result = 0;
+
+// Sum results from separate threads into result
+#pragma omp parallel reduction(+ \
+                               : result)
+    {
+        // Used to store the result of (M \otimes 1_d) * V
+        std::complex<double> term;
+
+        // Avoid unecessary multiply ops
+        size_t mat_col_offset;
+
+// Parallelise over d - likely to be a large value
+#pragma omp for
+        // OpenMP on Windows needs to use signed int for the index
+        // Use long int as muon and 9 spin 7/2 will use max integer value
+        for (long int k = 0; k < d; ++k) {
+            for (unsigned int i = 0; i < M_dim; ++i) {
+                mat_col_offset = i * M_dim;
+                term = {0, 0};
+                for (unsigned int j = 0; j < M_dim; ++j)
+                    // M[i, j] -> M[i * operator_dim + j] as column major
+                    term += M_ptr[mat_col_offset + j] * V_ptr[j * d + k];
+                // Sum current result of full product
+                result += (std::conj(V_ptr[i * d + k]) * term).real();
+            }
+        }
+    }
```

Comment:
This is more of a question based on my unfamiliarity with the langauge. It seems "wrong" to me for the outer look to be over `d` given that (in Python at least) I'd do the innermost part as `M[i,j] * V[j*d : j*d + k]`, i.e. accessing the `k` adjacent elements of V at the same time rather than iterating `d` times at the outermost level.

I assume that here what really matters (given the comment) we need `d` to be on the outside as it offers the best opportunity to parallelise, and given we (have to?) access every element individually anyway even the idea of slicing an array being more desirable than a loop is meaningless here?

Reply:
Yes exactly, there is no array slicing here and so the order doesn't matter, the array access is already the fastest it can be (up to calculating the indices) and we want the other loops inside the one that's parallelised to avoid unnecessary overhead by creating & deleting more threads.